cmake/compiler_settings.cmake

```diff
@@ -67,6 +67,22 @@ else ()
   message(WARNING "Spectre 1 mitigations NOT supported")
 endif ()
 
+# Apply Control Flow Guard if available.
+set(CONTROL_FLOW_GUARD_FLAGS /guard:cf)
+set(CONTROL_FLOW_GUARD_SHARED_LINKER_FLAGS /guard:cf /DYNAMICBASE)
+check_c_compiler_flag("${CONTROL_FLOW_GUARD_FLAGS}" CONTROL_FLOW_GUARD_FLAGS_C_SUPPORTED)
+check_cxx_compiler_flag("${CONTROL_FLOW_GUARD_FLAGS}"  CONTROL_FLOW_GUARD_FLAGS_CXX_FLAGS_SUPPORTED)
+if (CONTROL_FLOW_GUARD_FLAGS_C_SUPPORTED AND CONTROL_FLOW_GUARD_FLAGS_CXX_FLAGS_SUPPORTED)
+  message(STATUS "Control Flow GUARD Supported")
+  # We set this variable to indicate the flags are supported. It is
+  # empty otherwise.
+  set(CONTROL_FLOW_GUARD_FLAGS ${CONTROL_FLOW_GUARD_FLAGS})
+  set(CMAKE_SHARED_LINKER_FLAGS ${CONTROL_FLOW_GUARD_SHARED_LINKER_FLAGS})
+  add_compile_options(${CONTROL_FLOW_GUARD_FLAGS})
+else ()
+  message(WARNING "Control Flow Guard NOT supported")
```

Comment:
I think this will emit even on Linux. And while it's true there, it may not be relevant.

Reply:
There is an equivalent set of flags for GCC 8/Clang 7 that we should enable on the Linux side:
```
-fcf-protection=full
```
These are forward-looking features based on Intel's CET features due after the Icelake gen and should no-op if the CPU features are not available. A couple of caveats:

* I believe we're only on GCC7 today still, so this potentially breaks unless CMake can detect feature support and optionally enable. We should test this.
* There may be a perf impact to using `full` that we should validate with a quick test ... I would run something that doesn't do a lot of enclave transitions (i.e. don't use the thread tests) because that's likely to mask what I would expect to be minimal perf overheads.